contrib/zenith/file_cache.c

```diff
@@ -0,0 +1,345 @@
+ */
+#define CHUNK_SIZE 128 /* 1Mb chunk */
+
+typedef struct FileCacheEntry
+{
+	BufferTag key;
+	uint32    offset;
+	uint32    access_count;
+	uint32    bitmap[CHUNK_SIZE/32];
+	/* LRU list */
+	struct FileCacheEntry* next;
+	struct FileCacheEntry* prev;
+} FileCacheEntry;
+
+typedef struct FileCacheControl
+{
+	uint32 size; /* size of cache file in chunks */
+	FileCacheEntry lru;
+} FileCacheControl;
+
+static HTAB* lfc_hash;
+static int   lfc_desc;
+static LWLockId lfc_lock;
+static int   lfc_size;
+static char* lfc_path;
+static  FileCacheControl* lfc_ctl;
+static shmem_startup_hook_type prev_shmem_startup_hook;
+
+/*
+ * Prune L2 list
+ */
+static void
+lfc_prune(FileCacheEntry* entry)
+{
+	entry->next = entry->prev = entry;
+}
+
+/*
+ * Unlink from LRU list
+ */
+static void
+lfc_unlink(FileCacheEntry* entry)
+{
+	entry->next->prev = entry->prev;
+	entry->prev->next = entry->next;
+}
+
+/*
+ * Link to the head of LRU list
+ */
+static void
+lfc_link(FileCacheEntry* after, FileCacheEntry* entry)
+{
+	entry->next = after->next;
+	entry->prev = after;
+	after->next->prev = entry;
+	after->next = entry;
+}
+
+/*
+ * Check if list is empty
+ */
+static bool
+lfc_is_empty(FileCacheEntry* entry)
+{
+	return entry->next == entry;
+}
+
+static void
+lfc_shmem_startup(void)
+{
+	bool found;
+	static HASHCTL info;
+
+	if (prev_shmem_startup_hook)
+	{
+		prev_shmem_startup_hook();
+	}
+
+	LWLockAcquire(AddinShmemInitLock, LW_EXCLUSIVE);
+
+	lfc_ctl = (FileCacheControl*)ShmemInitStruct("lfc", sizeof(FileCacheControl), &found);
+	if (!found)
+	{
+		lfc_lock = (LWLockId)GetNamedLWLockTranche("lfc_lock");
+		info.keysize = sizeof(BufferTag);
+		info.entrysize = sizeof(FileCacheEntry);
+		lfc_hash = ShmemInitHash("lfc_hash",
+								 /* lfc_size+1 because we add new element to hash table before eviction of victim */
+								 lfc_size+1, lfc_size+1,
+								 &info,
+								 HASH_ELEM | HASH_BLOBS);
+		lfc_ctl->size = 0;
+		lfc_prune(&lfc_ctl->lru);
+
+		/* Remove file cache on restart */
+		(void)unlink(lfc_path);
+	}
+	LWLockRelease(AddinShmemInitLock);
+}
+
+void
+lfc_init(void)
+{
+	/*
+	 * In order to create our shared memory area, we have to be loaded via
+	 * shared_preload_libraries.
+	 */
+	if (!process_shared_preload_libraries_in_progress)
+		elog(ERROR, "Zenith module should be loaded via shared_preload_libraries");
+
+	DefineCustomIntVariable("zenith.file_cache_size",
+							"Maximal size of zenith local file cache (chunks)",
+							NULL,
+							&lfc_size,
+							0, /* disabled by default */
+							0,
+							INT_MAX,
+							PGC_POSTMASTER,
+							0,
+							NULL,
+							NULL,
+							NULL);
+
+	DefineCustomStringVariable("zenith.file_cache_path",
+							   "Path to local file cache (can be raw device)",
+							   NULL,
+							   &lfc_path,
+							   "file.cache",
+							   PGC_POSTMASTER,
+							   0,
+							   NULL,
+							   NULL,
+							   NULL);
+
+	if (lfc_size == 0)
```

Comment:
maybe also check for an unset lfc_path?

Reply:
`lfc_path` can be set to NULL (it has default value). Only to empty string.